src/json_utils.cpp

```diff
@@ -0,0 +1,109 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025 Datadog, Inc.
+
+#include <cstddef>
+#include <rapidjson/document.h>
+#include <rapidjson/error/error.h>
+#include <string_view>
+#include <tuple>
+#include <vector>
+
+#include "ddwaf.h"
+#include "json_utils.hpp"
+#include "utils.hpp"
+
+namespace ddwaf {
+
+ddwaf_object json_to_object(std::string_view json)
+{
+    // Safety limit, potentially not necessary as if we can decode and parse
+    // we should have a manageable json structure as other limits are already
+    // in place.
+    constexpr std::size_t max_depth = object_limits::default_max_container_depth;
+
+    auto to_object = [](const rapidjson::Value &doc) -> ddwaf_object {
+        ddwaf_object object;
+        if (doc.IsBool()) {
+            ddwaf_object_bool(&object, doc.GetBool());
+        } else if (doc.IsInt64()) {
+            ddwaf_object_signed(&object, doc.GetInt64());
+        } else if (doc.IsUint64()) {
+            ddwaf_object_unsigned(&object, doc.GetUint64());
+        } else if (doc.IsDouble()) {
+            ddwaf_object_float(&object, doc.GetDouble());
+        } else if (doc.IsString()) {
+            const std::string_view str = doc.GetString();
+            ddwaf_object_stringl(&object, str.data(), str.size());
+        } else if (doc.IsNull()) {
+            ddwaf_object_null(&object);
+        } else if (doc.IsObject()) {
+            ddwaf_object_map(&object);
+        } else if (doc.IsArray()) {
+            ddwaf_object_array(&object);
+        } else {
+            ddwaf_object_invalid(&object);
+        }
+        return object;
+    };
+
+    rapidjson::Document doc;
+    const rapidjson::ParseResult result = doc.Parse(json.data(), json.size());
```

Comment:
You could use the streaming parser to avoid the intermediary structures, this is especially the case here because you're not even taking advantage of the fact that you know in advance the size of the containers (the realloc overhead happened already inside rapidjson code). See e.g. https://github.com/DataDog/nginx-datadog/blob/master/src/security/body_parse/body_json.cpp#L89

Reply:
Good point, I'll have a look thanks

Reply:
> because you're not even taking advantage of the fact that you know in advance the size of the containers

In fairness, this will not be an issue in v2 since the new abstractions allow for preallocation.

Reply:
Done.